### ISSUE TITLE
Review (lab1) by Omar Ormachea (s304811)

### ISSUE THREAD

**Issue by oormacheah** [Marco-Masera/comp_int_2022#2]
# Lab 1 review by Omar Ormachea
First of all, I want to say that I'm impressed that this code was made from scratch without using the given code snippets and even with a recursive approach for the DFS. Aside from that, as Luca said in his review, it is very difficult to follow the logic of the program without comments or a more explanatory README file.

After debugging I managed to understand a bit more how the programs work. 

I followed a similar approach as your A* implementation, which as you know is very memory inefficient since all discovered nodes are being stored in memory at each node processing. The usage of a tuple as "priority" value for the heuristic function is very clever.

 Some minor issues I could spot:
-  On the class `List`:
   - `self.transformedSet = set(content)` is an unnecessary cast since the `content` parameter is always passed as a `set` already.
- On the class `List`:
  - I didn't quite understand what this line does and the reason behind its unusual format (I apologize if it's some advanced OOP style that I don't know): 
`list = list = [l.intersectTo_Copy(self.remainingElems) for index, l in enumerate(lists_) if index > skip]` 
Only thing I could say is that I would avoid the name "list" for a variable.

Overall, great job!

**Comment by Marco-Masera** [comment]
Hi. Thanks for the kind words.
I definitely will add comments and a more explanatory readme next time!
About the two points:
-set(originalSet) creates a copy of the original set instead of a
reference. It's not a cast but a function that returns a new set.
-Absolutely calling the variable "list" wasn't a great idea, I can see it
can be confusing
-The second of code you quoted iterates on a list of sets and for each one
calls the intersectTo_copy() function, which returns a new set. It
basically builds the list of sets for the given node. The index<cut serves
to avoid copying sets that won't be needed for the given node; this I
should have put in the readme: it's basically a rule to avoid generating
different nodes with different permutations of the same list of sets. If a
node has N sets, its i child will surely use set i, and the branch
sprouting from it cannot use any set j with j<i.

Il Dom 23 Ott 2022, 12:44 Omar Ormachea Hermoza ***@***.***>
ha scritto:

**Comment by oormacheah** [comment]
It's clear now, thanks. I used a slightly different approach for tackling the problem of avoiding permutations of same lists, but your approach makes total sense. And I apologize for my misunderstanding on the cast to set. Have a good one!
